File: clients/Go/model/attack_action.go
package model

import "io"
import . "aicup2020/stream"

type AttackAction struct {
    Target *int32
    AutoAttack *AutoAttack
}
func NewAttackAction(target *int32, autoAttack *AutoAttack) AttackAction {
    return AttackAction {
        Target: target,
        AutoAttack: autoAttack,
    }
}
func ReadAttackAction(reader io.Reader) AttackAction {
    result := AttackAction {}
    if ReadBool(reader) {
        var TargetValue int32
        TargetValue = ReadInt32(reader)
        result.Target = &TargetValue
    } else {
        result.Target = nil
    }
    if ReadBool(reader) {
        var AutoAttackValue AutoAttack
        AutoAttackValue = ReadAutoAttack(reader)
        result.AutoAttack = &AutoAttackValue
    } else {
        result.AutoAttack = nil
    }
    return result
}
func (value AttackAction) Write(writer io.Writer) {
    if value.Target == nil {
        WriteBool(writer, false)
    } else {
        WriteBool(writer, true)
        WriteInt32(writer, (*value.Target))
    }
    if value.AutoAttack == nil {
        WriteBool(writer, false)
    } else {
        WriteBool(writer, true)
        (*value.AutoAttack).Write(writer)
    }
}

File: clients/Python/model/attack_action.py
from .auto_attack import AutoAttack
class AttackAction:
    def __init__(self, target, auto_attack):
        self.target = target
        self.auto_attack = auto_attack
    @staticmethod
    def read_from(stream):
        if stream.read_bool():
            target = stream.read_int()
        else:
            target = None
        if stream.read_bool():
            auto_attack = AutoAttack.read_from(stream)
        else:
            auto_attack = None
        return AttackAction(target, auto_attack)
    def write_to(self, stream):
        if self.target is None:
            stream.write_bool(False)
        else:
            stream.write_bool(True)
            stream.write_int(self.target)
        if self.auto_attack is None:
            stream.write_bool(False)
        else:
            stream.write_bool(True)
            self.auto_attack.write_to(stream)
    def __repr__(self):
        return "AttackAction(" + \
            repr(self.target) + "," + \
            repr(self.auto_attack) + \
            ")"
